src/NLog/MessageTemplates/ValueFormatter.cs

```diff
@@ -47,21 +46,13 @@ namespace NLog.MessageTemplates
     /// </summary>
     internal class ValueFormatter : IValueFormatter
     {
-        public static IValueFormatter Instance
-        {
-            get => _instance ?? (_instance = new ValueFormatter(null)); //todo fix
-            set => _instance = value ?? new ValueFormatter(null);
-        }
-        private static IValueFormatter _instance;
         private static readonly IEqualityComparer<object> _referenceEqualsComparer = SingleItemOptimizedHashSet<object>.ReferenceEqualityComparer.Default;
 
-        private readonly IServiceRepository _serviceRepository;
-        private IJsonConverter JsonConverter => _jsonConverter ?? (_jsonConverter = _serviceRepository?.ResolveService<IJsonConverter>());
-        private IJsonConverter _jsonConverter;
+        private readonly IJsonConverter _jsonConverter;
 
-        internal ValueFormatter([NotNull] IServiceRepository serviceRepository)
+        public ValueFormatter([NotNull] IJsonConverter jsonConverter)
         {
-            _serviceRepository = serviceRepository ?? throw new ArgumentNullException(nameof(serviceRepository));
+            _jsonConverter = jsonConverter ?? throw new ArgumentNullException(nameof(jsonConverter));
```

Comment:
The late resolve of IJsonConverter was actually on purpose, so allow late override.

Reply:
With di there is no late resolve? Then we need a factory

Reply:
Thought DI was all about late resolve.